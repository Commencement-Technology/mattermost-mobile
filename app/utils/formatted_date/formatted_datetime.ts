// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import moment from 'moment';
import mtz from 'moment-timezone';

import type {FormatDateOptions} from 'react-intl';

export const formattedDateTime_moment = (date = new Date(), locale = 'en-us', format = 'MMM DD, YYYY', timezone?: string | UserTimezone | null): string => {
    moment.locale(locale);
    let formattedDate = mtz(date).format(format);
    if (timezone) {
        let zone: string;
        if (typeof timezone === 'object') {
            zone = timezone.useAutomaticTimezone ? timezone.automaticTimezone : timezone.manualTimezone;
        } else {
            zone = timezone;
        }
        formattedDate = mtz.tz(date, zone).format(format);
    }

    return formattedDate;
};

export const formattedDateTime_intl = (date = new Date(), locale = 'en-us', options?: FormatDateOptions): string => {
    return (new Intl.DateTimeFormat(locale, options)).format(date);
};
