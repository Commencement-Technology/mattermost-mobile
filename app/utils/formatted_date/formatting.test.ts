// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import languages from '@app/i18n/languages';

import {formattedDateTime_moment, formattedDateTime_intl} from './formatted_datetime';

import type {FormatDateOptions} from 'react-intl';

const MOMENT_FORMATS = [
    'MMM DD HH:MM A',
    'MMM DD YYYY HH:MM A',
    'dddd',
    'MMM DD',
    'MMM DD, YYYY',
];

const INTL_FORMATS: { [key: string]: FormatDateOptions } = {
    'MMM DD HH:MM A': {
        month: 'short',
        day: '2-digit',
        hour: '2-digit',
        minute: '2-digit',
    },
    'MMM DD YYYY HH:MM A': {
        year: 'numeric',
        month: 'short',
        day: '2-digit',
        hour: '2-digit',
        minute: '2-digit',
    },
    dddd: {
        weekday: 'long',
    },
    'MMM DD': {
        month: 'short',
        day: '2-digit',
    },
    'MMM DD, YYYY': {
        month: 'short',
        day: '2-digit',
        year: 'numeric',
    },
};

describe.each(MOMENT_FORMATS)('formattedDateTime', (format: string) => {
    const theResult: string[] = [];

    const theLanguages = Object.keys(languages);

    afterAll(() => {
        expect(theResult).toMatchSnapshot();
    });

    it.each(theLanguages)(`format: ${format}`, (locale) => {
        const theDate = new Date('2021-09-02T00:00:00Z');
        const formatMoment = formattedDateTime_moment(theDate, locale, format);
        const formatIntl = formattedDateTime_intl(theDate, locale, INTL_FORMATS[format]);

        theResult.push(`${locale} :: ${formatMoment}   ===>   ${formatIntl}`);
    });
});
