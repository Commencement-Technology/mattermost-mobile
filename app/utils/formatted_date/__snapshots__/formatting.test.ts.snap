// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formattedDateTime format: MMM DD 1`] = `
[
  "bg :: сеп 01   ===>   1 сеп",
  "de :: Sep. 01   ===>   01. Sept.",
  "en :: Sep 01   ===>   Sep 01",
  "en-AU :: Sep 01   ===>   01 Sept",
  "es :: sep. 01   ===>   01 sept",
  "fa :: سپتامبر ۰۱   ===>   ۱۰ شهریور",
  "fr :: sept. 01   ===>   01 sept.",
  "hu :: szept. 01   ===>   szept. 01.",
  "it :: set 01   ===>   01 set",
  "ja :: 9月 01   ===>   9月01日",
  "ko :: 9월 01   ===>   9월 01일",
  "nl :: sep. 01   ===>   01 sep",
  "pl :: wrz 01   ===>   01 wrz",
  "pt-BR :: set 01   ===>   01 de set.",
  "ro :: sept. 01   ===>   01 sept.",
  "ru :: сент. 01   ===>   01 сент.",
  "sv :: sep 01   ===>   01 sep.",
  "tr :: Eyl 01   ===>   01 Eyl",
  "uk :: вер 01   ===>   01 вер.",
  "vi :: Thg 09 01   ===>   01 thg 9",
  "zh-CN :: 9月 01   ===>   9月01日",
  "zh-TW :: 9月 01   ===>   9月01日",
]
`;

exports[`formattedDateTime format: MMM DD HH:MM A 1`] = `
[
  "bg :: сеп 01 18:09 PM   ===>   1 сеп, 18:00 ч.",
  "de :: Sep. 01 18:09 PM   ===>   01. Sept., 18:00",
  "en :: Sep 01 18:09 PM   ===>   Sep 01, 06:00 PM",
  "en-AU :: Sep 01 18:09 PM   ===>   01 Sept, 06:00 pm",
  "es :: sep. 01 18:09 PM   ===>   01 sept, 18:00",
  "fa :: سپتامبر ۰۱ ۱۸:۰۹ بعد از ظهر   ===>   ۱۰ شهریور، ۱۸:۰۰",
  "fr :: sept. 01 18:09 PM   ===>   01 sept., 18:00",
  "hu :: szept. 01 18:09 DU   ===>   szept. 01. 18:00",
  "it :: set 01 18:09 PM   ===>   01 set, 18:00",
  "ja :: 9月 01 18:09 午後   ===>   9月01日 18:00",
  "ko :: 9월 01 18:09 오후   ===>   9월 01일 오후 06:00",
  "nl :: sep. 01 18:09 PM   ===>   01 sep 18:00",
  "pl :: wrz 01 18:09 PM   ===>   01 wrz, 18:00",
  "pt-BR :: set 01 18:09 PM   ===>   01 de set., 18:00",
  "ro :: sept. 01 18:09 PM   ===>   01 sept., 18:00",
  "ru :: сент. 01 18:09 вечера   ===>   01 сент., 18:00",
  "sv :: sep 01 18:09 PM   ===>   01 sep. 18:00",
  "tr :: Eyl 01 18:09 ÖS   ===>   01 Eyl 18:00",
  "uk :: вер 01 18:09 вечора   ===>   01 вер., 18:00",
  "vi :: Thg 09 01 18:09 CH   ===>   18:00 01 thg 9",
  "zh-CN :: 9月 01 18:09 晚上   ===>   9月01日 18:00",
  "zh-TW :: 9月 01 18:09 晚上   ===>   9月01日 下午06:00",
]
`;

exports[`formattedDateTime format: MMM DD YYYY HH:MM A 1`] = `
[
  "bg :: сеп 01 2021 18:09 PM   ===>   01.09.2021 г., 18:00 ч.",
  "de :: Sep. 01 2021 18:09 PM   ===>   01. Sept. 2021, 18:00",
  "en :: Sep 01 2021 18:09 PM   ===>   Sep 01, 2021, 06:00 PM",
  "en-AU :: Sep 01 2021 18:09 PM   ===>   01 Sept 2021, 06:00 pm",
  "es :: sep. 01 2021 18:09 PM   ===>   01 sept 2021, 18:00",
  "fa :: سپتامبر ۰۱ ۲۰۲۱ ۱۸:۰۹ بعد از ظهر   ===>   ۱۰ شهریور ۱۴۰۰، ۱۸:۰۰",
  "fr :: sept. 01 2021 18:09 PM   ===>   01 sept. 2021, 18:00",
  "hu :: szept. 01 2021 18:09 DU   ===>   2021. szept. 01. 18:00",
  "it :: set 01 2021 18:09 PM   ===>   01 set 2021, 18:00",
  "ja :: 9月 01 2021 18:09 午後   ===>   2021年9月01日 18:00",
  "ko :: 9월 01 2021 18:09 오후   ===>   2021년 9월 01일 오후 06:00",
  "nl :: sep. 01 2021 18:09 PM   ===>   01 sep 2021 18:00",
  "pl :: wrz 01 2021 18:09 PM   ===>   01 wrz 2021, 18:00",
  "pt-BR :: set 01 2021 18:09 PM   ===>   01 de set. de 2021, 18:00",
  "ro :: sept. 01 2021 18:09 PM   ===>   01 sept. 2021, 18:00",
  "ru :: сент. 01 2021 18:09 вечера   ===>   01 сент. 2021 г., 18:00",
  "sv :: sep 01 2021 18:09 PM   ===>   01 sep. 2021 18:00",
  "tr :: Eyl 01 2021 18:09 ÖS   ===>   01 Eyl 2021 18:00",
  "uk :: вер 01 2021 18:09 вечора   ===>   01 вер. 2021 р., 18:00",
  "vi :: Thg 09 01 2021 18:09 CH   ===>   18:00 01 thg 9, 2021",
  "zh-CN :: 9月 01 2021 18:09 晚上   ===>   2021年9月01日 18:00",
  "zh-TW :: 9月 01 2021 18:09 晚上   ===>   2021年9月01日 下午06:00",
]
`;

exports[`formattedDateTime format: MMM DD, YYYY 1`] = `
[
  "bg :: сеп 01, 2021   ===>   01.09.2021 г.",
  "de :: Sep. 01, 2021   ===>   01. Sept. 2021",
  "en :: Sep 01, 2021   ===>   Sep 01, 2021",
  "en-AU :: Sep 01, 2021   ===>   01 Sept 2021",
  "es :: sep. 01, 2021   ===>   01 sept 2021",
  "fa :: سپتامبر ۰۱، ۲۰۲۱   ===>   ۱۰ شهریور ۱۴۰۰",
  "fr :: sept. 01, 2021   ===>   01 sept. 2021",
  "hu :: szept. 01, 2021   ===>   2021. szept. 01.",
  "it :: set 01, 2021   ===>   01 set 2021",
  "ja :: 9月 01, 2021   ===>   2021年9月01日",
  "ko :: 9월 01, 2021   ===>   2021년 9월 01일",
  "nl :: sep. 01, 2021   ===>   01 sep 2021",
  "pl :: wrz 01, 2021   ===>   01 wrz 2021",
  "pt-BR :: set 01, 2021   ===>   01 de set. de 2021",
  "ro :: sept. 01, 2021   ===>   01 sept. 2021",
  "ru :: сент. 01, 2021   ===>   01 сент. 2021 г.",
  "sv :: sep 01, 2021   ===>   01 sep. 2021",
  "tr :: Eyl 01, 2021   ===>   01 Eyl 2021",
  "uk :: вер 01, 2021   ===>   01 вер. 2021 р.",
  "vi :: Thg 09 01, 2021   ===>   01 thg 9, 2021",
  "zh-CN :: 9月 01, 2021   ===>   2021年9月01日",
  "zh-TW :: 9月 01, 2021   ===>   2021年9月01日",
]
`;

exports[`formattedDateTime format: dddd 1`] = `
[
  "bg :: сряда   ===>   сряда",
  "de :: Mittwoch   ===>   Mittwoch",
  "en :: Wednesday   ===>   Wednesday",
  "en-AU :: Wednesday   ===>   Wednesday",
  "es :: miércoles   ===>   miércoles",
  "fa :: چهارشنبه   ===>   چهارشنبه",
  "fr :: mercredi   ===>   mercredi",
  "hu :: szerda   ===>   szerda",
  "it :: mercoledì   ===>   mercoledì",
  "ja :: 水曜日   ===>   水曜日",
  "ko :: 수요일   ===>   수요일",
  "nl :: woensdag   ===>   woensdag",
  "pl :: środa   ===>   środa",
  "pt-BR :: quarta-feira   ===>   quarta-feira",
  "ro :: miercuri   ===>   miercuri",
  "ru :: среда   ===>   среда",
  "sv :: onsdag   ===>   onsdag",
  "tr :: Çarşamba   ===>   Çarşamba",
  "uk :: середа   ===>   середа",
  "vi :: thứ tư   ===>   Thứ Tư",
  "zh-CN :: 星期三   ===>   星期三",
  "zh-TW :: 星期三   ===>   星期三",
]
`;
